examples/stm32/xively_demo_stm32f4xx_nucleo_wifi/Src/ntp.c

```diff
@@ -0,0 +1,301 @@
+        goto terminate;
+    }
+    printf("\r\n\tUDP Connection [OK] Assigned socket ID: %d", *sock_id);
+
+    /* Send SNTP request */
+    wifi_retval = sntp_send_request(*sock_id);
+    if( wifi_retval != WiFi_MODULE_SUCCESS )
+    {
+        printf("\r\n\tSNTP send_request [FAIL] Retval: %d", wifi_retval);
+        retval = SNTP_REQUEST_FAILURE;
+        goto terminate;
+    }
+
+    /* Await SNTP Response */
+    printf("\r\n\tAwaiting server response");
+    if( sntp_await_response(*sock_id) < 0 )
+    {
+        printf("\r\n\tSNTP response [FAIL] Response timed out");
+        retval = SNTP_TIMEOUT;
+        goto terminate;
+    }
+
+    if( *sock_id != last_sntp_response->socket_id )
+    {
+        printf("\r\n\tSocket ID assertion [FAIL] Bug in ntp.c?");
+        retval = SNTP_INTERNAL_ERROR;
+        goto terminate;
+    }
+
+    /* Parse server response */
+    printf("\r\n\tParsing and converting SNTP response");
+    sntp_current_time = sntp_parse_response(last_sntp_response->response);
+    if( sntp_current_time < 0 )
+    {
+        printf("\r\n\tSNTP Parsing and conversion [FAIL]");
+        retval = SNTP_PARSER_ERROR;
+    }
+    else
+    {
+        printf("\r\n\tReceived epoch time: %ld", sntp_current_time);
+        retval = SNTP_SUCCESS;
+    }
+
+terminate:
+    /* free() SNTP Response */
+    if( NULL != last_sntp_response )
+    {
+        sntp_free_response(last_sntp_response);
+        last_sntp_response = NULL;
+    }
+
+    /* Close socket */
+    if( *sock_id > 0 )
+    {
+        sntp_disconnect(*sock_id);
+    }
+    return retval;
+}
+
+/**
+   * @brief
+   * @param  sock_id we're interested in
+   * @retval SNTP_SUCCESS (0) or SNTP_TIMEOUT(-1)
+   */
+sntp_status_t sntp_await_response( uint8_t sock_id )
+{
+    int32_t sntp_timeout_ms = SNTP_TIMEOUT_MS;
+    const int32_t timeout_step = 250;
+    while( NULL == last_sntp_response )
+    {
+        if( sntp_timeout_ms <= 0 )
+        {
+            return SNTP_TIMEOUT;
+        }
+        printf(".");
+        HAL_Delay(timeout_step);
+        sntp_timeout_ms -= timeout_step;
+    }
+    return SNTP_SUCCESS;
+}
+
+/**
+   * @brief  This function shall be called from ind_wifi_socket_data_received(),
+   *         when the sock_id is the one we got from sntp_get_datetime()
+   * @param  See description for ind_wifi_socket_data_received
+   * @retval None
+   */
+void sntp_socket_data_callback(uint8_t sock_id, uint8_t* data_ptr,
+                               uint32_t message_size, uint32_t chunk_size)
+{
+    /* Verify we got an SNTP response, not just protocol data */
+    if( (SNTP_MSG_SIZE != message_size) || (SNTP_MSG_SIZE != chunk_size) )
+    {
+        printf("\r\n\tMessage isn't an SNTP respone. It will be ignored");
+        return;
+    }
+
+    /* Store SNTP response */
+    if( NULL != last_sntp_response )
+    {
+        sntp_free_response(last_sntp_response);
+    }
+    last_sntp_response = calloc(1, sizeof(last_sntp_response));
```

Comment:
Please use only ```malloc``` and ```memset``` as calloc may not be available on some platforms... yeah I know this is written for this platform but still.... if we just would like to reuse this code somewhere else it would be nice to have such option.